src/generic.jl

```diff
@@ -120,6 +120,35 @@ function _pairwise!(r::AbstractMatrix, metric::SemiMetric, a::AbstractMatrix)
     r
 end
 
+function _pairwise!(r::AbstractMatrix, metric::PreMetric,
+                    a::AbstractVector, b::AbstractVector=a)
+    na = length(a)
+    nb = length(b)
+    size(r) == (na, nb) || throw(DimensionMismatch("Incorrect size of r."))
+    @inbounds for (j, bj) = enumerate(b)
+        for (i, ai) = enumerate(a)
+            r[i, j] = metric(ai, bj)
+        end
+    end
+    r
+end
+
+function _pairwise!(r::AbstractMatrix, metric::SemiMetric, a::AbstractVector)
+    n = length(a)
+    size(r) == (n, n) || throw(DimensionMismatch("Incorrect size of r."))
+    @inbounds for j = 1:n
+        for i = (j + 1):n
+            r[i, j] = metric(a[i], a[j])
+        end
+        r[j, j] = 0
+        for i = 1:(j - 1)
+            r[i, j] = r[j, i]   # leveraging the symmetry of SemiMetric
+        end
```

Comment:
How about flipping the two loops? Then you would go through the array in the order of the elements.

Reply:
I used the same approach as for matrices as in line 107 but I can switch it again.